### ISSUE TITLE
Support MSYS2 MinGW-w64

### ISSUE THREAD

**Issue by ghost** [NazaraEngine/NazaraEngine#393]
xmake config --mode=releasedbg failed to install assimp:

=> install assimp v5.2.5 .. failed

```
C:/msys64/mingw64/include/c++/12.2.0/cmath:45:15: fatal error: math.h: No such file or directory
   45 | #include_next <math.h>
      |               ^~~~~~~~
compilation terminated.
  ${yellow}> in ${bright}C:\msys64\tmp\.xmake\230301\_9FBCAB64B619410AA5B53DAE28D84EC2.cpp
if you want to get more verbose errors, please see:
  -> C:\Users\iahuntn\AppData\Local\.xmake\cache\packages\2303\a\assimp\v5.2.5\installdir.failed\logs\install.txt
error: install failed!
```

Update: `xmake` installed with `pacman -S mingw-w64-x86_64-xmake`

**Comment by SirLynix** [comment]
Hi,

This is caused by https://github.com/msys2/MINGW-packages/issues/10761.
The only fix I'm aware of is to disable the usage of `-isystem` and rely on `-I` instead.

You can do this by disabling the `package.include_external_headers:n` policy.
`xmake config --mode=releasedbg --policies=package.include_external_headers:n`

this is what's done on the [MinGW CI](https://github.com/NazaraEngine/NazaraEngine/blob/main/.github/workflows/msys2-build.yml#L97)

**Comment by ghost** [comment]
> Hi,
> 
> This is caused by [msys2/MINGW-packages#10761](https://github.com/msys2/MINGW-packages/issues/10761). The only fix I'm aware of is to disable the usage of `-isystem` and rely on `-I` instead.
> 
> You can do this by disabling the `package.include_external_headers:n` policy. `xmake config --mode=releasedbg --policies=package.include_external_headers:n`
> 
> this is what's done on the [MinGW CI](https://github.com/NazaraEngine/NazaraEngine/blob/main/.github/workflows/msys2-build.yml#L97)

It doesn't work!

```
  => install assimp v5.2.5 .. failed

C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0x9ea5): undefined reference to `unzOpen'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0x9fc6): undefined reference to `unzGoToFirstFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0xa02e): undefined reference to `unzGetCurrentFileInfo'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0xa0c8): undefined reference to `unzOpenCurrentFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0xa0f8): undefined reference to `unzReadCurrentFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0xa129): undefined reference to `unzCloseCurrentFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0xa192): undefined reference to `unzGoToNextFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(IFCLoader.cpp.obj):IFCLoader.cpp:(.text+0xa1a3): undefined reference to `unzClose'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0x818): undefined reference to `unzClose'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0xbdb): undefined reference to `unzGoToFilePos'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0xbeb): undefined reference to `unzOpenCurrentFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0xc7a): undefined reference to `unzReadCurrentFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0xdf3): undefined reference to `unzOpen2'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0xe32): undefined reference to `unzClose'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0x11cc): undefined reference to `unzOpen2'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0x11dc): undefined reference to `unzClose'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\iahuntn\AppData\Local\.xmake\packages\a\assimp\v5.2.5\315bab8ab3e04a7ebfc148df500488da\lib/libassimp.a(ZipArchiveIOSystem.cpp.obj):ZipArchiveIOSystem.cpp:(.text+0x12df): undefined reference to `unzOpen2'
if you want to get more verbose errors, please see:
  -> C:\Users\iahuntn\AppData\Local\.xmake\cache\packages\2303\a\assimp\v5.2.5\installdir.failed\logs\install.txt
error: install failed!
```

**Comment by SirLynix** [comment]
This is a whole different error.

Could you try to run `xmake config --mode=releasedbg --policies=package.include_external_headers:n -vD` and send the full log here to help debug it?

Also, if you don't need assimp for now you can disable it with `--assimp=n`

**Comment by ghost** [comment]
`xmake` has native support for msys2, my solution is use `pacman` packages to speed up the build process and also as a workaround for this problem. Sometimes, the software doesn't compile out of the box and needs the msys2 maintainer to patch it. Most of the time, mixing packaging systems will result in errors like this. I still remember my misery with `libarchive` and `pkgsrc` (`pkgsrc` uses it own patched `libarchive`, but on the bootstrap process it links with the system's `libarchive` even though I passed option to force it to prefer it own version `libarchive`). Yeah, this is on FreeBSD but I learned the lesson (I have never able to have `pkgsrc` works on FreeBSD because of this problem).

Long story short, when it ask you `please input: y (y/n/m)`, choose `m` and manually force it to use `pacman` packages instead of building it own.

**Comment by SirLynix** [comment]
It works well on the MinGW/msys2 CI, I'd like to know why it's failing for you.

**Comment by SirLynix** [comment]
Closed since user deleted their account and lacking info to reproduce.

**Comment by waruqi** [comment]
I have improved it. https://github.com/xmake-io/xmake/commit/35dd6cbd7c94be4d27eacb9d88e66ee7ce11b473

**Comment by SirLynix** [comment]
Thank you! I will test this
